include/nbl/video/IGPUCommandPool.h

```diff
@@ -4,38 +4,803 @@
+{
+public:
+    CDrawIndirectCountCmd(core::smart_refctd_ptr<const video::IGPUBuffer>&& buffer, core::smart_refctd_ptr<const video::IGPUBuffer>&& countBuffer)
+        : m_buffer(std::move(buffer)) , m_countBuffer(std::move(countBuffer))
+    {}
+
+private:
+    core::smart_refctd_ptr<const IGPUBuffer> m_buffer;
+    core::smart_refctd_ptr<const IGPUBuffer> m_countBuffer;
+};
+
+class IGPUCommandPool::CDrawIndexedIndirectCountCmd : public IGPUCommandPool::IFixedSizeCommand<CDrawIndexedIndirectCountCmd>
+{
+public:
+    CDrawIndexedIndirectCountCmd(core::smart_refctd_ptr<const video::IGPUBuffer>&& buffer, core::smart_refctd_ptr<const video::IGPUBuffer>&& countBuffer)
+        : m_buffer(std::move(buffer)), m_countBuffer(std::move(countBuffer))
+    {}
+
+private:
+    core::smart_refctd_ptr<const IGPUBuffer> m_buffer;
+    core::smart_refctd_ptr<const IGPUBuffer> m_countBuffer;
+};
+
+class IGPUCommandPool::CBeginRenderPassCmd : public IGPUCommandPool::IFixedSizeCommand<CBeginRenderPassCmd>
+{
+public:
+    CBeginRenderPassCmd(core::smart_refctd_ptr<const video::IGPURenderpass>&& renderpass, core::smart_refctd_ptr<const video::IGPUFramebuffer>&& framebuffer)
+        : m_renderpass(std::move(renderpass)), m_framebuffer(std::move(framebuffer))
+    {}
+
+private:
+    core::smart_refctd_ptr<const video::IGPURenderpass> m_renderpass;
+    core::smart_refctd_ptr<const video::IGPUFramebuffer> m_framebuffer;
+};
+
+class IGPUCommandPool::CPipelineBarrierCmd : public IGPUCommandPool::IFixedSizeCommand<CPipelineBarrierCmd>
+{
+public:
+    CPipelineBarrierCmd(const uint32_t bufferCount, const core::smart_refctd_ptr<const IGPUBuffer>* buffers, const uint32_t imageCount, const core::smart_refctd_ptr<const IGPUImage>* images)
+    {
+        m_barrierResources = core::make_refctd_dynamic_array<core::smart_refctd_dynamic_array<core::smart_refctd_ptr<const core::IReferenceCounted>>>(imageCount + bufferCount);
+
+        uint32_t k = 0;
+
+        for (auto i = 0; i < bufferCount; ++i)
+            m_barrierResources->begin()[k++] = buffers[i];
+
+        for (auto i = 0; i < imageCount; ++i)
+            m_barrierResources->begin()[k++] = images[i];
+    }
+
+private:
+    core::smart_refctd_dynamic_array<core::smart_refctd_ptr<const core::IReferenceCounted>> m_barrierResources;
+};
+
+class IGPUCommandPool::CBindDescriptorSetsCmd : public IGPUCommandPool::IFixedSizeCommand<CBindDescriptorSetsCmd>
+{
+public:
+    CBindDescriptorSetsCmd(core::smart_refctd_ptr<const IGPUPipelineLayout>&& pipelineLayout, const uint32_t setCount, const core::smart_refctd_ptr<const IGPUDescriptorSet>* sets)
+        : m_layout(std::move(pipelineLayout))
+    {
+        m_sets = core::make_refctd_dynamic_array<core::smart_refctd_dynamic_array< core::smart_refctd_ptr<const IGPUDescriptorSet>>>(setCount);
+
+        for (auto i = 0; i < setCount; ++i)
+            m_sets->begin()[i] = sets[i];
```

Comment:
there's only ever 4, its an engine imposed maximum

Reply:
do not use `smart_refctd_dynamic_array`, use `std::array`

Reply:
Resolved.